src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -5,61 +5,90 @@ import {
 	AbstractInstantiationError,
 	NotImplementedError,
 } from '@tvkitchen/base-errors'
+import { TSDemuxer } from 'ts-demuxer'
+
 import logger from '%src/lib/logger'
 import kafka from '%src/lib/kafka'
+import { tsToSeconds } from '%src/lib/utils/mpegts'
 
 class AbstractIngestionEngine {
 	constructor() {
 		if (this.constructor === AbstractIngestionEngine) {
 			throw new AbstractInstantiationError(this.constructor.name)
 		}
 		this.producer = kafka.producer()
+		this.mpegtsDemuxer = new TSDemuxer(this.processDemuxedPacket)
+		this.ffmpegProcess = null
+		this.currentPts = 0
 	}
 
 	/**
-	 * Ingests data into the TV Kitchen pipeline.
+	 * Ingests a mpegts payload into the TV Kitchen pipeline.
 	 *
-	 * @param  {Buffer} data The buffer of data to be ingested.
-	 * @return {Boolean}     Whether or not the buffer was successfully ingested.
+	 * @param  {Payload} data The payload to be ingested.
+	 * @return {Boolean}      Whether or not the payload was successfully ingested.
 	 */
-	ingestData = async (data) => {
-		const payload = new Payload()
-		payload.data = data
+	ingestPayload = async (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('ingestPayload must be passed a Payload object')
+		}
 		await this.producer.send({
-			topic: dataTypes.STREAM.DATA,
+			topic: dataTypes.STREAM.CONTAINER,
 			messages: [{
-				value: data,
+				value: Payload.serialize(payload),
 			}],
 		})
 		return true
 	}
 
 	/**
-	 * Continuously ingest of a stream of data.
+	 * Updates ingestion state based on the latest mpegts packet data as
+	 * parsed by our demuxer.
 	 *
-	 * @param  {ReadableStream} readableStream The stream of data to be recorded.
-	 * @return {ReadableStream}                The input stream (for chaining)
+	 * @param  {Packet} packet A Packet object, as defined by the TSDemuxer project
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
+	processDemuxedPacket = (packet) => {
+		this.currentPts = packet.pts
+	}
+
+	/**
+	 * Process a new chunk of data from an mpegts stream.
+	 * This data is used to update local state and ultimately pipe to kafka.
+	 *
+	 * @param  {Buffer} mpegtsData The latest arbitrary chunk of buffered mpegts data.
+	 */
+	processMpegtsData = (mpegtsData) => {
+		this.ffmpegProcess.stdout.pause()
```

Comment:
Hey there, I haven't spent enough time with this to know for sure, but using an on data event and pausing the stream here may not the most clean and efficient way to accomplish throttling. It's possible there's some limiting constraint here that I'm not seeing, or experience that informs this choice, so waves hands ignore me if so, but stream backpressure might be a better way to do throttling. If you want, I can take a look at what's involved in that, and whether the result would be in fact cleaner/more efficient either for this PR or a subsequent one after creating an issue.

Reply:
Oooo nice catch -- and you actually just exposed a nasty bug!

So to your solid throttle point: I believe we no longer need (or want) to throttle in the way I was toying with during the R&D stuff -- where I was trying to make sure ffprobe and ffmpeg would process at roughly the same speed.

The *original* intent of the `.pause` and `.resume` was in order to prevent a race condition where kafka messages were sometimes being recorded out of order.

For instance:

```
- Data A arrives (new thread: A)
- A: Write to Kafka
- Data B arrives (new thread: B)
- B: Write to Kafka
- B: Write complete
- A: write complete
```

By pausing, we ensure that `Data B arrives` won't happen until after `A: Write complete`

However, this code *does not accomplish that goal*.  Specifically:

`this.ingestPayload(...)`

should be

`await this.ingestPayload()`

---

So, with that context I *think* we still need to pause / resume, as something of a mutex, or maybe what we really should do is use a *real* mutex plus the backpressure solution you're describing?

Did anything I just wrote actually make sense or did I accidentally just write 70 lines of buzzwords.

Reply:
Side note: I have no idea how to reliably test for race conditions...

Reply:
You go to the track and drive there.

Okay, so I think that's in part because the events continue to fire even if a prior event hasn't completed. It's async with no way to wait. A through or write stream pattern would handle this more gracefully by preserving order without a mutex or pausing/resuming.

Reply:
Serious answer about how to test: you might mock out the Kafka ingest, then block on that until the second one fires.

Reply:
Oh yes! I'll add a test like that.

And yes that's exactly it -- unless I paused, events would keep firing.

Reply:
Okay, expanding on this. What you've got now is essentially:
```js
ffmpeg.stdout.on('data',data => {
    ffmpeg.stdout.pause()
    ingestPayload(data)
    ffmpegs.stdout.resume()
})

inputStream.on('data',data => {
    inputStream.pause()
    ffmpegs.stdin.write(data)
    inputStream.resume()
})
```

What I'm talking about looks something like this:

```js
const processStream = Writable({
    write: (data, enc, done) => {
        ingestPayload(data)
            .then(response => done())
            .catch(done)
    }
});

ffmpeg.stdout.pipe(processStream)

inputStream.pipe(ffmpeg.stdin)
```

I took a stab at applying this to the PR branch in a fork, you can find that [here](https://github.com/tvkitchen/tv-kitchen/compare/45-parse-progress...chriszs:45-parse-progress). It looks this there's a way to suggest code changes to a set of lines in a PR comment, but this seems a little larger than that was designed for. Anyway, it's there as a reference/basis for future work if you want it.

One question I have is whether ts-demuxer's process method calls the method that updates the timestamp synchronously. If it doesn't, there may be some mismatch in timing. I read its [code](https://github.com/gliese1337/HLS.js/blob/master/demuxer/src/index.ts), but couldn't be 100% sure.

Reply:
Thinking about this further this morning, something like this is possible and a little clearer about what each step does:
```js
import { Writable, Transform } from 'stream'

const processStream = Transform({
	objectMode: true,
	transform: (data, enc, done) => {
		TSdemuxer.process(data)
		done(null, new Payload{ payload: data, pts: this.currentPts })
	},
});

const ingestStream = Writable({
	objectMode: true,
	write: (data, enc, done) => {
        	ingestPayload(data)
            		.then(() => done())
            		.catch(done)
	},
});

inputStream.pipe(ffmpeg.stdin)

ffmpeg.stdout.pipe(processStream).pipe(ingestStream)
```

Could alternatively use pipeline syntax:
```js
import { pipeline } from 'stream'

inputStream.pipe(ffmpeg.stdin)

pipeline(
	ffmpeg.stdout,
	processStream,
	ingestStream,
	() => console.log('done'),
);
```
Callback style shown but it's promisfiable.

Reply:
Oh, this is so nice -- implementing now (you're already a co-author on this commit but I kinda wanna add your name twice)

Reply:
This is now done -- and boy, testing this approach was not easy!